pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,272 @@
+package alertmanager
+
+import (
+	"context"
+	"hash/fnv"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/httpgrpc"
+	"github.com/weaveworks/common/httpgrpc/server"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/alertmanager/alertmanagerpb"
+	"github.com/cortexproject/cortex/pkg/ring"
+	"github.com/cortexproject/cortex/pkg/ring/client"
+	"github.com/cortexproject/cortex/pkg/tenant"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// Distributor forwards requests to individual alertmanagers.
+type Distributor struct {
+	services.Service
+
+	cfg              *MultitenantAlertmanagerConfig
+	requestsInFlight sync.WaitGroup
+
+	alertmanagerRing        ring.ReadRing
+	alertmanagerClientsPool ClientsPool
+
+	logger log.Logger
+
+	receivedRequests *prometheus.CounterVec
+	amSends          *prometheus.CounterVec
+	amSendFailures   *prometheus.CounterVec
+}
+
+// NewDistributor constructs a new Distributor
+func NewDistributor(cfg *MultitenantAlertmanagerConfig, alertmanagersRing *ring.Ring, alertmanagerClientsPool ClientsPool, logger log.Logger, reg prometheus.Registerer) (d *Distributor, err error) {
+	if alertmanagerClientsPool == nil {
+		alertmanagerClientsPool = newAlertmanagerClientsPool(client.NewRingServiceDiscovery(alertmanagersRing), cfg.AlertmanagerClient, logger, reg)
+	}
+
+	d = &Distributor{
+		cfg:                     cfg,
+		logger:                  logger,
+		alertmanagerRing:        alertmanagersRing,
+		alertmanagerClientsPool: alertmanagerClientsPool,
+	}
+
+	d.receivedRequests = promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
```

Comment:
The naming of these metrics is a bit ambiguous. The prefix is `cortex_alertmanager_distributor_` so it looks to be from the alertmanager distributor's perspective. If so, what you call "received request" is actually "sent request" and viceversa. The distributor is the client, the alertmanager is the server.

Other than this, few more comments:
1. Do we need "receivedRequests" to be per user?
2. What's the "ingester" label in "amSends" and "amSendFailures"? Looks wrong to me
3. Since we have to rename these metrics, I would suggest you to take a look at current cortex request/response metric names to keep these specular (to find them, you can just open Prometheus and type "cortexrequest")

Reply:
These request metrics would be taken care by the grpc server, so none of these metrics are required.